### PR TITLE
fix #1054 Distinct allows for arbitrary store and duplicate check

### LINKED ISSUE #1054: Distinct should allow use of arbitrary data structures
### Expected behavior

It'd be nice to be able to pass arbitrary ways to test distinctiveness to the `Flux.distinct` method. For example, in my project I want to pass a Guava cache to `distinct` to allow time-based eviction of elements when working with an infinite stream of events - but the method signature demands a `Supplier<Collection>`. I got around it with `Collections.newSetFromMap(cache.asMap())`, but not all caches might expose read/write `Map` views (for example, Spring caches do not).

One way to resolve this would be to add a method overload with two new parameters: some distinct datastructure store `S` with `Supplier<S>`, and some way to test an element `V` for distinctiveness `BiPredicate<S, V>`. With this, my usecase would look something like: 
```java
flux.distinct(element -> element.getKey(),
              () -> getCache(),
              (cache, element) -> cache.getIfPresent(element) == null));
```

### Actual behavior

See above.

### Steps to reproduce

Not relevant, feature request.

### Reactor Core version

3.1.2.RELEASE

### JVM version (e.g. `java -version`)

java version "1.8.0_152"
Java(TM) SE Runtime Environment (build 1.8.0_152-b16)
Java HotSpot(TM) 64-Bit Server VM (build 25.152-b16, mixed mode)

### OS version (e.g. `uname -a`)

Darwin [HOSTNAME REMOVED] 16.7.0 Darwin Kernel Version 16.7.0: Mon Nov 13 21:56:25 PST 2017; root:xnu-3789.72.11~1/RELEASE_X86_64 x86_64

### PR DESCRIPTION
**Pull request by simonbasle**
This commit removes the assumption that the distinct operator performs
its check via storing in a `Collection` and the `add` method. Instead,
it lets the user define an arbitrary store and a `BiPredicate` that
takes that store and the extracted key in order to determine if a value
has been encountered before.

A cleanup consumer is also used on termination, although like in the
former `Collection#clear`-based implementation it is assumed to not
raise an Exception.

### CODE REVIEW

**Comment by simonbasle** [comment]
@smaldini what do you think about this one?

I will try to protect the `cleanupCallback` calls, but I had an operator test mysteriously failing after my initial attempt at this, so for now I assume that `cleanupCallback` doesn't throw (like it was the case with the `Collection.clear()` beforehand).

(PS: operator test abstraction is very very convoluted and could benefit from a complete do-over 🤕 I wonder if there is an avenue to explore in JUnit test case generation, like the collection test generators from Guava)

File: reactor-core/src/main/java/reactor/core/publisher/Flux.java [hunk 1]
```diff
@@ -3497,11 +3497,43 @@ public final T blockLast(Duration timeout) {
 	public final <V, C extends Collection<? super V>> Flux<T> distinct(
 			Function<? super T, ? extends V> keySelector,
 			Supplier<C> distinctCollectionSupplier) {
+		return this.distinct(keySelector, distinctCollectionSupplier, Collection::add, Collection::clear);
```
Comment (smaldini) [inline reactor-core/src/main/java/reactor/core/publisher/Flux.java]: could store static variable in FluxDistinct

**Comment by smaldini** [comment]
I desperately agree with the test comment @simonbasle :(

File: reactor-core/src/main/java/reactor/core/publisher/Flux.java (hunk 1 above)
Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/Flux.java]: I don't think method references benefit from storing in a static variable

**Comment by simonbasle** [comment]
@smaldini opened #1061 for the test improvement

**Pull request merged**